File: my-reader-mobile/src/services/core/content.ts
import type { Library } from "@my-reader/tools/types/library"
import { withLocalLibraryCalibreRoot } from "../fs/local-library-content"
import {
  contentClearCoverThumbnailCache,
  contentDeleteCoverThumbnailCache,
  contentDeleteFileState,
  contentFinalizeDownloadedFile,
  contentGetFileState,
  contentListCoverThumbnailCache,
  contentListFileStates,
  contentListReadingFormats,
  contentMarkFileRemoteOnly,
  contentSetReadingFormat,
  contentUpsertCoverThumbnailCache,
  contentUpsertFileState,
  type BookCoverThumbnailCache as CoreBookCoverThumbnailCache,
  type BookCoverThumbnailCachePatch as CoreBookCoverThumbnailCachePatch,
  type DownloadedFile as CoreDownloadedFile,
  type FileState as CoreFileState,
} from "my-reader-core"
import { librarySidecarRootUri } from "../fs/library-paths"
import { toNativeFilesystemPath } from "../fs/path"
import {
  invalidateBookReadingFormat,
  invalidateFileStates,
} from "../query/invalidate-table"

export type FileState = Omit<
  CoreFileState,
  "localBlake3" | "localSize" | "localMtime"
> & {
  localBlake3: string | null
  localSize: number | null
  localMtime: number | null
}

export type FileStateUpdate = {
  localState: FileState["localState"]
  localBlake3?: string | null
  localSize?: number | null
  localMtime?: number | null
}

export type DownloadedFile = CoreDownloadedFile

export type BookCoverThumbnailCache = CoreBookCoverThumbnailCache

export type BookCoverThumbnailCachePatch = CoreBookCoverThumbnailCachePatch

function sidecarRootPath(library: Library): string {
  return toNativeFilesystemPath(librarySidecarRootUri(library))
}

function fileStateFromCore(state: CoreFileState): FileState {
  return {
    ...state,
    localBlake3: state.localBlake3 ?? null,
    localSize: state.localSize ?? null,
    localMtime: state.localMtime ?? null,
  }
}

export async function listBookReadingFormats(
  library: Library,
): Promise<Record<string, string>> {
  const formats = await withLocalLibraryCalibreRoot(library, (libraryRootUri) =>
    contentListReadingFormats(
      sidecarRootPath(library),
      toNativeFilesystemPath(libraryRootUri),
    ),
  )
  return Object.fromEntries(
    formats.map(({ bookId, format }) => [bookId, format]),
  )
}

export async function setBookReadingFormat(
  library: Library,
  bookId: number,
  format: string | null,
): Promise<void> {
  await withLocalLibraryCalibreRoot(library, (libraryRootUri) =>
    contentSetReadingFormat(
      sidecarRootPath(library),
      toNativeFilesystemPath(libraryRootUri),
      bookId,
      format ?? undefined,
    ),
  )
  await invalidateBookReadingFormat(library.id)
}

export function getFileState(
  library: Library,
  path: string,
): Promise<FileState | null> {
  return contentGetFileState(sidecarRootPath(library), path).then((state) =>
    state ? fileStateFromCore(state) : null,
  )
}

export async function listFileStates(library: Library): Promise<FileState[]> {
  return (await contentListFileStates(sidecarRootPath(library))).map(
    fileStateFromCore,
  )
}

export async function upsertFileState(
  library: Library,
  path: string,
  update: FileStateUpdate,
): Promise<void> {
  await contentUpsertFileState(sidecarRootPath(library), path, {
    localState: update.localState,
    localBlake3: update.localBlake3 ?? undefined,
    localSize: update.localSize ?? undefined,
    localMtime: update.localMtime ?? undefined,
  })
  await invalidateFileStates(library.id)
}

export async function deleteFileState(
  library: Library,
  path: string,
): Promise<void> {
  await contentDeleteFileState(sidecarRootPath(library), path)
  await invalidateFileStates(library.id)
}

export async function finalizeDownloadedFile(
  library: Library,
  relativePath: string,
  localFileUri: string,
): Promise<DownloadedFile> {
  const downloaded = await contentFinalizeDownloadedFile(
    sidecarRootPath(library),
    relativePath,
    toNativeFilesystemPath(localFileUri),
  )
  await invalidateFileStates(library.id)
  return downloaded
}

export async function markFileRemoteOnly(
  library: Library,
  relativePath: string,
): Promise<void> {
  await contentMarkFileRemoteOnly(sidecarRootPath(library), relativePath)
  await invalidateFileStates(library.id)
}

export function listBookCoverThumbnailCache(
  library: Library,
  input: {
    thumbnailVersion: string
    widthPx: number
    heightPx: number
  },
): Promise<BookCoverThumbnailCache[]> {
  return contentListCoverThumbnailCache(
    sidecarRootPath(library),
    input.thumbnailVersion,
    input.widthPx,
    input.heightPx,
  )
}

export function upsertBookCoverThumbnailCache(
  library: Library,
  patch: BookCoverThumbnailCachePatch,
): Promise<void> {
  return contentUpsertCoverThumbnailCache(sidecarRootPath(library), patch)
}

export function deleteBookCoverThumbnailCache(
  library: Library,
  input: {
    bookId: number
    thumbnailVersion: string
    widthPx: number
    heightPx: number
  },
): Promise<void> {
  return contentDeleteCoverThumbnailCache(
    sidecarRootPath(library),
    input.bookId,
    input.thumbnailVersion,
    input.widthPx,
    input.heightPx,
  )
}

export function clearBookCoverThumbnailCache(library: Library): Promise<void> {
  return contentClearCoverThumbnailCache(sidecarRootPath(library))
}

File: my-reader-mobile/modules/my-reader-core/rust/src/content.rs
use std::path::Path;

use my_reader_core::api::content::ContentService;

use crate::{
    types::{
        required_i64, BookCoverThumbnailCache, BookCoverThumbnailCachePatch, DownloadedFile,
        FileState, FileStateUpdate, ReadingFormat,
    },
    CoreFfiError,
};

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_list_reading_formats(
    sidecar_root_path: String,
    library_root_path: String,
) -> Result<Vec<ReadingFormat>, CoreFfiError> {
    Ok(ContentService::list_reading_formats(
        Path::new(&sidecar_root_path),
        Path::new(&library_root_path),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(|(book_id, format)| ReadingFormat { book_id, format })
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_set_reading_format(
    sidecar_root_path: String,
    library_root_path: String,
    book_id: f64,
    format: Option<String>,
) -> Result<(), CoreFfiError> {
    ContentService::set_reading_format(
        Path::new(&sidecar_root_path),
        Path::new(&library_root_path),
        required_i64(book_id, "bookId")?,
        format.as_deref(),
    )
    .await
    .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_get_file_state(
    sidecar_root_path: String,
    path: String,
) -> Result<Option<FileState>, CoreFfiError> {
    Ok(
        ContentService::get_file_state(Path::new(&sidecar_root_path), &path)
            .await
            .map_err(CoreFfiError::from_core)?
            .map(Into::into),
    )
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_list_file_states(
    sidecar_root_path: String,
) -> Result<Vec<FileState>, CoreFfiError> {
    Ok(
        ContentService::list_file_states(Path::new(&sidecar_root_path))
            .await
            .map_err(CoreFfiError::from_core)?
            .into_iter()
            .map(Into::into)
            .collect(),
    )
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_upsert_file_state(
    sidecar_root_path: String,
    path: String,
    update: FileStateUpdate,
) -> Result<(), CoreFfiError> {
    ContentService::upsert_file_state(Path::new(&sidecar_root_path), &path, update.try_into()?)
        .await
        .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_delete_file_state(
    sidecar_root_path: String,
    path: String,
) -> Result<(), CoreFfiError> {
    ContentService::delete_file_state(Path::new(&sidecar_root_path), &path)
        .await
        .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_finalize_downloaded_file(
    sidecar_root_path: String,
    relative_path: String,
    local_path: String,
) -> Result<DownloadedFile, CoreFfiError> {
    Ok(ContentService::finalize_downloaded_file(
        Path::new(&sidecar_root_path),
        &relative_path,
        Path::new(&local_path),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_mark_file_remote_only(
    sidecar_root_path: String,
    relative_path: String,
) -> Result<(), CoreFfiError> {
    ContentService::mark_file_remote_only(Path::new(&sidecar_root_path), &relative_path)
        .await
        .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_list_cover_thumbnail_cache(
    sidecar_root_path: String,
    thumbnail_version: String,
    width_px: f64,
    height_px: f64,
) -> Result<Vec<BookCoverThumbnailCache>, CoreFfiError> {
    Ok(ContentService::list_cover_thumbnail_cache(
        Path::new(&sidecar_root_path),
        &thumbnail_version,
        required_i64(width_px, "widthPx")?,
        required_i64(height_px, "heightPx")?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_upsert_cover_thumbnail_cache(
    sidecar_root_path: String,
    patch: BookCoverThumbnailCachePatch,
) -> Result<(), CoreFfiError> {
    ContentService::upsert_cover_thumbnail_cache(Path::new(&sidecar_root_path), patch.try_into()?)
        .await
        .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_delete_cover_thumbnail_cache(
    sidecar_root_path: String,
    book_id: f64,
    thumbnail_version: String,
    width_px: f64,
    height_px: f64,
) -> Result<(), CoreFfiError> {
    ContentService::delete_cover_thumbnail_cache(
        Path::new(&sidecar_root_path),
        required_i64(book_id, "bookId")?,
        &thumbnail_version,
        required_i64(width_px, "widthPx")?,
        required_i64(height_px, "heightPx")?,
    )
    .await
    .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn content_clear_cover_thumbnail_cache(
    sidecar_root_path: String,
) -> Result<(), CoreFfiError> {
    ContentService::clear_cover_thumbnail_cache(Path::new(&sidecar_root_path))
        .await
        .map_err(CoreFfiError::from_core)
}
